File: ios/NativeCore/Sources/NativeCore/PlaceReference.swift
import Foundation
import CoreFoundation

public struct Coordinate: Equatable {
    public let latitude, longitude: Double
    public init(latitude: Double, longitude: Double) throws {
        guard latitude.isFinite, longitude.isFinite, (-90...90).contains(latitude), (-180...180).contains(longitude) else { throw PlaceError.invalidCoordinate }
        self.latitude = latitude; self.longitude = longitude
    }
    public static func from(item: [String: Any]) -> Coordinate? {
        guard let lat = item["lat"] as? NSNumber, let lng = item["lng"] as? NSNumber,
              CFGetTypeID(lat) != CFBooleanGetTypeID(), CFGetTypeID(lng) != CFBooleanGetTypeID() else { return nil }
        return try? Coordinate(latitude: lat.doubleValue, longitude: lng.doubleValue)
    }
}
public enum PlaceReference: Equatable { case manual(Coordinate), google(placeID: String), existing, unlocated }
public struct PlaceSelection {
    public let reference: PlaceReference
    public let originalItem: [String: Any]?
    public let sourceTripID: String?
    public let sourceItemKey: ItemKey?
    public let sourceSavedID: String?
    public init(reference: PlaceReference, originalItem: [String: Any]? = nil, sourceTripID: String? = nil, sourceItemKey: ItemKey? = nil, sourceSavedID: String? = nil) {
        self.reference = reference; self.originalItem = originalItem; self.sourceTripID = sourceTripID; self.sourceItemKey = sourceItemKey; self.sourceSavedID = sourceSavedID
    }
    public func validated() throws -> PlaceSelection {
        if case .google(let id) = reference {
            let trimmed = id.trimmingCharacters(in: .whitespacesAndNewlines)
            guard (1...1024).contains(trimmed.unicodeScalars.count) else { throw PlaceError.invalidReference }
            return .init(reference: .google(placeID: trimmed))
        }
        if case .existing = reference {
            guard let originalItem, let id = originalItem["id"] else { throw PlaceError.invalidReference }
            if let sourceSavedID {
                guard sourceTripID == nil, !sourceSavedID.isEmpty, id as? String == sourceSavedID else { throw PlaceError.invalidReference }
            } else {
                guard let sourceTripID, !sourceTripID.isEmpty, let sourceItemKey, try ItemKey(json: id) == sourceItemKey else { throw PlaceError.invalidReference }
            }
        }
        return self
    }
    func sourceKey(creationID: String) throws -> String {
        let value = try validated()
        switch value.reference {
        case .google(let id): return "google:" + id
        case .manual, .unlocated: return "manual:" + creationID
        case .existing:
            let parts = value.sourceSavedID.map { ["saved", $0] } ?? ["existing", value.sourceTripID!, value.sourceItemKey!.token]
            return "existing:" + TripBackup.hash(try JSONSerialization.data(withJSONObject: parts, options: [.withoutEscapingSlashes]))
        }
    }
    func durablePayload(id: String, draft: PlaceDraft) throws -> [String: Any] {
        guard UUID(uuidString: id) != nil else { throw ScheduleError.invalidID }
        let value = try validated()
        var raw: [String: Any] = value.reference == .existing ? value.originalItem! : [:]
        try draft.validate(original: raw)
        raw["id"] = id
        for (key, field) in draft.fields { raw[key] = field }
        switch value.reference {
        case .manual(let coordinate): raw["lat"] = coordinate.latitude; raw["lng"] = coordinate.longitude
        case .google(let placeID): raw["placeId"] = placeID; raw["nativePlaceSource"] = "google"
        case .existing, .unlocated: break
        }
        return raw
    }
}
public struct ResolvedPlace {
    public let placeID: String
    public let coordinate: Coordinate
    public let displayName, address: String
    public let attribution: [String]
    public let fetchedAt: Date
    public init(placeID: String, coordinate: Coordinate, displayName: String, address: String, attribution: [String], fetchedAt: Date) {
        self.placeID = placeID; self.coordinate = coordinate; self.displayName = displayName; self.address = address; self.attribution = attribution; self.fetchedAt = fetchedAt
    }
}
public enum PlaceError: LocalizedError {
    case invalidCoordinate, invalidReference, conflictingOperation
    public var errorDescription: String? {
        switch self {
        case .invalidCoordinate: return "위도(-90~90)와 경도(-180~180)를 함께 확인해주세요."
        case .invalidReference: return "장소 정보를 확인하지 못했습니다. 다시 선택해주세요."
        case .conflictingOperation: return "같은 저장 요청의 내용이 변경되었습니다. 다시 확인해주세요."
        }
    }
}

File: android/nativepreview/src/main/java/com/travelplaner/nativepreview/domain/PlaceReference.kt
package com.travelplaner.nativepreview.domain
import kotlinx.serialization.json.*

data class Coordinate(val latitude: Double, val longitude: Double) {
    init { require(latitude.isFinite() && longitude.isFinite() && latitude in -90.0..90.0 && longitude in -180.0..180.0) { "위도(-90~90)와 경도(-180~180)를 함께 확인해주세요." } }
    companion object {
        fun from(item: JsonObject): Coordinate? {
            val lat = item["lat"] as? JsonPrimitive ?: return null
            val lng = item["lng"] as? JsonPrimitive ?: return null
            if (lat.isString || lng.isString) return null
            return runCatching { Coordinate(lat.double,lng.double) }.getOrNull()
        }
    }
}
sealed interface PlaceReference {
    data class Manual(val coordinate: Coordinate): PlaceReference
    data class Google(val placeID: String): PlaceReference
    data object Existing: PlaceReference
    data object Unlocated: PlaceReference
}
data class PlaceSelection(val reference: PlaceReference, val originalItem: JsonObject? = null, val sourceTripID: String? = null, val sourceItemKey: ItemKey? = null, val sourceSavedID: String? = null) {
    fun validated(): PlaceSelection {
        if (reference is PlaceReference.Google) {
            val id = reference.placeID.trim()
            require(id.codePointCount(0,id.length) in 1..1024) { "장소 정보를 다시 선택해주세요." }
            return PlaceSelection(PlaceReference.Google(id))
        }
        if (reference == PlaceReference.Existing) {
            require(originalItem != null)
            if (sourceSavedID != null) require(sourceTripID == null && sourceSavedID.isNotEmpty() && originalItem["id"] == JsonPrimitive(sourceSavedID))
            else require(!sourceTripID.isNullOrEmpty() && sourceItemKey != null && originalItem["id"]?.let(ItemKey::from) == sourceItemKey) { "원본 장소 정보를 확인하지 못했어요." }
        }
        return this
    }
    fun sourceKey(creationID: String): String {
        val value = validated()
        return when (val reference = value.reference) {
            is PlaceReference.Google -> "google:" + reference.placeID
            is PlaceReference.Manual, PlaceReference.Unlocated -> "manual:" + creationID
            PlaceReference.Existing -> {
                val parts = value.sourceSavedID?.let { listOf("saved",it) } ?: listOf("existing",value.sourceTripID!!,value.sourceItemKey!!.token)
                "existing:" + TripBackup.hash(JsonArray(parts.map(::JsonPrimitive)).toString().encodeToByteArray())
            }
        }
    }
    fun durablePayload(id: String,draft: PlaceDraft): JsonObject {
        require(runCatching { java.util.UUID.fromString(id).toString().equals(id,true) }.getOrDefault(false))
        val value = validated()
        val raw = if (value.reference == PlaceReference.Existing) value.originalItem!!.toMutableMap() else mutableMapOf()
        draft.validate(JsonObject(raw)); raw["id"] = JsonPrimitive(id)
        raw.putAll(draft.fields.mapValues { JsonPrimitive(it.value) })
        when(val ref = value.reference) {
            is PlaceReference.Manual -> { raw["lat"] = JsonPrimitive(ref.coordinate.latitude); raw["lng"] = JsonPrimitive(ref.coordinate.longitude) }
            is PlaceReference.Google -> { raw["placeId"] = JsonPrimitive(ref.placeID); raw["nativePlaceSource"] = JsonPrimitive("google") }
            PlaceReference.Existing, PlaceReference.Unlocated -> Unit
        }
        return JsonObject(raw)
    }
}
data class ResolvedPlace(val placeID: String, val coordinate: Coordinate, val displayName: String, val address: String, val attribution: List<String>, val fetchedAt: Long)
